packages/xpath/src/functions/xforms/node-set.ts

```diff
@@ -32,6 +34,221 @@ export const countNonEmpty = new NumberFunction(
+
+		// Iterate through rest of arguments, collecting pairs of:
+		//
+		// - `repeats`: **all** nodes matching the supplied node-set for the
+		//   `repeatN` argument in this pair
+		// - `position`: the resolved number value for the `indexN` (per spec)
+		//   argument at in this pair
+		//
+		// For **all `repeatN`/`indexN` pairs**, arguments are evaluated in the
+		// calling context (in typical XForms usage, this will be the context of the
+		// bound node). This is the core difference between this approach and the
+		// original in https://github.com/getodk/web-forms/pull/150. That
+		// understanding was clarified in review of that orignal effort, and is
+		// borne out by new tests exercising depth > 1, which demonstrate the same
+		// behavior in JavaRosa.
+		//
+		// Note: we start iterating here at index 1 so assertions related to
+		// positional argument index are clear.
+		for (let i = 1; i < args.length; i += 2) {
+			const repeatsArg = args[i];
+			const positionArg = args[i + 1];
+
+			assertArgument(i, repeatsArg);
+			assertArgument(i + 1, positionArg);
+
+			// Evaluate position first, because...
+			const position = positionArg.evaluate(context).toNumber();
+
+			// ... if any "index" (position) is `NaN`, we short-circuit. This is
+			// expected behavior because the equivalent `/data/repN[posN]/target`
+			// expression would do the same.
+			if (Number.isNaN(position)) {
+				return [];
+			}
+
+			// Reiterating the point made describing this loop for future clarity:
+			// this collects **all** of the nodes matching the `repeatN` expression.
+			// We filter them in a later step.
+			const repeats = evaluateArgumentToFilterableNodes(context, repeatsArg);
+
+			// No repeats = nothing to "index" = short circuit
+			if (repeats.length === 0) {
+				return [];
+			}
+
+			pairs.push({
+				repeats,
+				position,
+			});
+		}
+
+		// Sort the results of each `repeatN`/`indexN` pair, by containment order.
+		//
+		// Note: the `repeatN`/`indexN` pairs can be supplied in any order (this is
+		// consistent with behavior in JavaRosa, likely as a side effect of the
+		// function being implemented there by transforming the expression to its
+		// LocationPath equivalent).
+		pairs = pairs.sort(compareContainmentDepth);
+
+		// Resolve repeats at the specified/evaluated position, in document depth
+		// order by:
+		//
+		// 1. Filtering each set of repeats to include **only** the nodes contained
+		//    by the previously resolved repeat.
+		//
+		// 2. Selecting the repeat at the specified/evaluated position (of those
+		//    filtered in 1).
+		let repeatContextNode: ContextNode;
+
+		for (const pair of pairs) {
+			const { position } = pair;
+			const repeats = pair.repeats.filter((repeat) => {
+				return repeatContextNode?.contains(repeat) !== false;
+			});
+
+			// Select next repeat context at the current `repeatN`/`indexN` position.
+			//
+			// Note: despite terminology used in the spec, `indexN` is treated as
```

Comment:
This comment is meant to clarify the implementation, not the spec. Specifically it's intended to explain why the implementation:

1. uses `position` rather than `index` terminology (as "index" has a meaning in JS/TS, and "position" is the XPath semantic terminology used for 1-based indexing throughout `@getodk/xpath`)
2. accesses `repeats[position - 1]` rather than `repeats[position]` (or `repeats[index]` if it had used the spec terminology).

I have frequently found it helpful to call out XPath semantics with the language used in the XPath spec, both in naming and comment prose. Especially so when I revisit implementation details in `@getodk/xpath` after the fact, as the package has generally been stable enough that it can be easy for me to forget various XPath semantic nuances I haven't thought about for a while.

Reply:
Ahhh, ok. As long as the comment makes sense to you it's ok! My brain keeps tripping up on the "despite"